### PR TITLE
feature: cli support add kill command

### LINKED ISSUE #2917: [feature] pouch cli supports kill function
## Why you need it?

We should implement 'pouch kill' to kill one or more running containers. 

## How it could be?

Example:
# pouch ps
Name     ID       Status          Created        Image            Runtime
3bc260   3bc260   Up 10 minutes   49 years ago   busybox:latest   runc

# pouch kill 3bc260
3bc260

# pouch ps
Name   ID   Status   Created   Image   Runtime

## Other related information

Usage
pouch kill [OPTIONS] CONTAINER [CONTAINER...]

Options
Name, shorthand	Default	Description
--signal , -s	KILL	Signal to send to the container

Extended description
You can kill a container using the container’s ID, ID-prefix, or name.

### PR DESCRIPTION
**Pull request by lang710**
Signed-off-by: Lang Chi <21860405@zju.edu.cn>

### Ⅰ. Describe what this PR did
cli support add kill command

### Ⅱ. Does this pull request fix one issue?

fixes #2917 

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
Added

### Ⅳ. Describe how to verify it
root@compatibility:~# pouch run -d busybox top
df0487e6cc386f9340d27e62ab8a3797e5ccccb5d199bbebb87a0155aa69a84f
root@compatibility:~# pouch kill df0487e6cc386f9340d27e62ab8a3797e5ccccb5d199bbebb87a0155aa69a84f
df0487e6cc386f9340d27e62ab8a3797e5ccccb5d199bbebb87a0155aa69a84f
root@compatibility:~# pouch run -d busybox top
75d8d902f78315b065c41e81a2a9967f7a3edf891c36bd50f36ab3b1691d9c73
root@compatibility:~# pouch kill -s 0 75d8d902f78315b065c41e81a2a9967f7a3edf891c36bd50f36ab3b1691d9c73
Error: {"message":"Invalid signal: 0"}

### CODE REVIEW

File: daemon/mgr/container.go [hunk 1]
```diff
@@ -68,6 +68,9 @@ type ContainerMgr interface {
 	// List returns the list of containers.
 	List(ctx context.Context, option *ContainerListOption) ([]*Container, error)
 
+	// Kill one or more running containers
+	Kill(ctx context.Context, name string, signal uint64) (err error)
```
Comment (ZYecho) [inline daemon/mgr/container.go]: the desc here I think is kill one running container， wdyt？

File: daemon/mgr/container_kill.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
+package mgr
+
+import (
+	"context"
+	"fmt"
+	"github.com/sirupsen/logrus"
```
Comment (ZYecho) [inline daemon/mgr/container_kill.go]: please format these imports

File: daemon/mgr/container_kill.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	"strings"
+	"syscall"
+)
+
+type errNoSuchProcess struct {
```
Comment (ZYecho) [inline daemon/mgr/container_kill.go]: could we put this into errdef package？

File: daemon/mgr/container_kill.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		}
+	}
+	return nil
+}
+
+func (mgr *ContainerManager) killDeadProcess(ctx context.Context, c *Container, signal int) error {
+	err := mgr.killWithSignal(ctx, c, signal)
+	if err == syscall.ESRCH {
+		e := errNoSuchProcess{c.State.Pid, signal}
+		logrus.Debug(e)
```
Comment (ZYecho) [inline daemon/mgr/container_kill.go]: please add more msg to this err

File: daemon/mgr/container_kill.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	if !c.IsRunning() {
+		return fmt.Errorf("Container %s is not running", c.ID)
+	}
+
+	if err := mgr.killDeadProcess(ctx, c, int(syscall.SIGKILL)); err != nil {
+		if isErrNoSuchProcess(err) {
+			return nil
+		}
+
+		if c.IsRunning() {
```
Comment (ZYecho) [inline daemon/mgr/container_kill.go]: why `killDeadProcess` don't lock the container here？

File: daemon/mgr/container_kill.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	if !c.State.Running {
+		return fmt.Errorf("Container %s is not running", c.ID)
+	}
+
+	msg, err := mgr.Client.DestroyContainer(ctx, c.ID, 10)
+	if err != nil {
+		err = fmt.Errorf("Cannot kill container %s: %s", c.ID, err)
+		if strings.Contains(err.Error(), "container not found") ||
+			strings.Contains(err.Error(), "no such process") {
+			logrus.Warnf("container kill failed because of 'container not found' or 'no such process': %s", err.Error())
```
Comment (ZYecho) [inline daemon/mgr/container_kill.go]: return err？

**Comment by ZYecho** [comment]
add more input：https://github.com/alibaba/pouch/issues/2733

File: daemon/mgr/container_kill.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+)
+
+// Kill kills a container.
+func (mgr *ContainerManager) Kill(ctx context.Context, name string, signal uint64) (err error) {
+	c, err := mgr.container(name)
+	if err != nil {
+		return err
+	}
+
+	if signal == 0 || syscall.Signal(signal) == syscall.SIGKILL {
```
[the remark this answers is not in the document]
Comment (lang710) [reply daemon/mgr/container_kill.go]: Thanks for your advice. I think that signal == 0 is redundant.

File: test/cli_kill_test.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/alibaba/pouch/test/command"
+	"github.com/alibaba/pouch/test/environment"
+	"github.com/go-check/check"
```
[the remark this answers is not in the document]
Comment (lang710) [reply test/cli_kill_test.go]: Thanks.

File: ctrd/container.go [hunk 9]
```diff
@@ -352,6 +352,46 @@ func (c *Client) recoverContainer(ctx context.Context, id string, io *containeri
...
+	}
+	defer c.lock.Unlock(containerID)
+
+	pack, err := c.watch.get(containerID)
+	if err != nil {
+		return err
+	}
+	// the caller need to execute the all hooks.
+	pack.l.Lock()
+	pack.skipStopHooks = true
```
[the remark this answers is not in the document]
Comment (lang710) [reply ctrd/container.go]: Actually I don't  know if it is necessary to add these things. But if you want to kill a container,  you should skip the stop hooks.

File: apis/server/container_bridge.go [hunk 10]
```diff
@@ -186,6 +187,24 @@ func (s *Server) getContainers(ctx context.Context, rw http.ResponseWriter, req
...
+
+	// If we have a signal, look at it. Otherwise, do nothing
+	if sigStr := req.FormValue("signal"); sigStr != "" {
+		var err error
+		if sig, err = utils.ParseSignal(sigStr); err != nil {
+			return err
+		}
+	}
+
+	if err := s.ContainerMgr.Kill(ctx, name, uint64(sig)); err != nil {
```
Comment (allencloud) [inline apis/server/container_bridge.go]: please just `return s.ContainerMgr.Kill(ctx, name, uint64(sig))` to simplify code.

File: daemon/mgr/container_kill.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+import (
+	"context"
+	"fmt"
+	"strings"
+	"syscall"
+
+	"github.com/pkg/errors"
+
+	"github.com/alibaba/pouch/pkg/errtypes"
+	"github.com/sirupsen/logrus"
```
Comment (allencloud) [inline daemon/mgr/container_kill.go]: please restruct the package import sequence.

File: daemon/mgr/container_kill.go [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+		if errtypes.IsNoSuchProcess(err) {
+			return nil
+		}
+
+		if c.IsRunning() {
+			return err
+		}
+	}
+
+	if pid := c.State.Pid; pid != 0 {
```
Comment (allencloud) [inline daemon/mgr/container_kill.go]: Please use `fast return` to reduce ident:
https://github.com/alibaba/pouch/blob/master/docs/contributions/code_styles.md#rule006---return-fast-to-indent-less

File: daemon/mgr/container_kill.go [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+	return err
+}
+
+func (mgr *ContainerManager) killWithSignal(ctx context.Context, c *Container, signal int) error {
+	logrus.Debugf("Sending %d to %s", signal, c.ID)
+	c.Lock()
+	defer c.Unlock()
+
+	if c.State.Paused {
+		return fmt.Errorf("Container %s is paused. Unpause the container before stopping", c.ID)
```
Comment (allencloud) [inline daemon/mgr/container_kill.go]: before killing?

File: docs/commandline/pouch_kill.md [hunk 14]
```diff
@@ -0,0 +1,49 @@
+## pouch kill
```
Comment (allencloud) [inline docs/commandline/pouch_kill.md]: Please remove this file. This file is auto-generated. https://github.com/alibaba/pouch/tree/master/docs#commandline
```
What's more, all commandline docs are auto generated via source code
```

File: pkg/utils/utils.go [hunk 15]
```diff
@@ -475,3 +550,20 @@ func ExtractIPAndPortFromAddresses(addresses []string) (string, string) {
 	}
 	return "", ""
 }
+
+// ParseSignal translates a string to a valid syscall signal.
+// It returns an error if the signal map doesn't include the given signal.
+func ParseSignal(rawSignal string) (syscall.Signal, error) {
```
Comment (allencloud) [inline pkg/utils/utils.go]: please add unit test for this function. Thanks.

**Comment by cxz66666** [comment]
ping @lang710 
Can you have time to update those questions? Or I will closed this PR. Thank you for your response!

**Comment by cxz66666** [comment]
Seem that author doesn't response for a long time, I open another [PR](https://github.com/alibaba/pouch/pull/3034) but some different details(don't skip hooks), so I will close PR.

**Pull request closed without merging**